# Prithvi Shetty 



In [206]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity='all'


## 1.Explore the data 

In [3]:
d=pd.read_csv('titanic.csv.bz2')

In [98]:
d.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [44]:
d['survived'].groupby([d.pclass]).mean()

pclass
1    0.619195
2    0.429603
3    0.255289
Name: survived, dtype: float64

In [104]:
b=["pclass" + i for i in d.pclass.unique().astype(str)]
b

['pclass1', 'pclass2', 'pclass3']

In [49]:
d['survived'].groupby([d.sex]).mean()

sex
female    0.727468
male      0.190985
Name: survived, dtype: float64

In [112]:
c=["Sex "+ i for i in d.sex.unique().astype(str)]
c

['Sex female', 'Sex male']

In [65]:
d['survived'].groupby([d.parch]).mean()

parch
0    0.335329
1    0.588235
2    0.504425
3    0.625000
4    0.166667
5    0.166667
6    0.000000
9    0.000000
Name: survived, dtype: float64

In [135]:
d['survived'].groupby([d.embarked]).mean()

embarked
C    0.555556
Q    0.357724
S    0.332604
Name: survived, dtype: float64

In [146]:
c2=list(['Port C','Port Q','Port S'])

In [51]:
d.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

In [92]:
a=d.parch.unique().astype(str)

In [93]:
a = ["parch" + i for i in a]

In [94]:
a

['parch0',
 'parch2',
 'parch1',
 'parch4',
 'parch3',
 'parch5',
 'parch6',
 'parch9']

In [73]:
from collections import OrderedDict

In [148]:
df1=pd.DataFrame(OrderedDict({'Variable': a ,'Survived':d['survived'].groupby([d.parch]).mean().values}))
df1

Variable  Survived
0   parch0  0.335329
1   parch2  0.588235
2   parch1  0.504425
3   parch4  0.625000
4   parch3  0.166667
5   parch5  0.166667
6   parch6  0.000000
7   parch9  0.000000

In [117]:
df2=pd.DataFrame(OrderedDict({'Variable': b ,'Survived':d['survived'].groupby([d.pclass]).mean().values}))
df2

Variable  Survived
0  pclass1  0.619195
1  pclass2  0.429603
2  pclass3  0.255289

In [118]:
df3=pd.DataFrame(OrderedDict({'Variable': c ,'Survived':d['survived'].groupby([d.sex]).mean().values}))
df3

Variable  Survived
0  Sex female  0.727468
1    Sex male  0.190985

In [160]:
df4=pd.DataFrame(OrderedDict({'Variable': c2 ,'Survived':d['survived'].groupby([d.embarked]).mean().values}))
df4

Variable  Survived
0   Port C  0.555556
1   Port Q  0.357724
2   Port S  0.332604

In [161]:
df=pd.concat([df2,df3,df1,df4])

In [162]:
df['Drowned']=1-df.Survived

In [168]:
df['Missing']=0

In [170]:
df.Missing.iloc[[13, 14, 15]] = 2

C:\Users\Prathik\Anaconda3\lib\site-packages\pandas\core\indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [341]:
df['Survival rate']=df.Survived*100

### Final summarized results by variable

In [342]:
df

Variable  Survived   Drowned  Missing  Survival rate
0      pclass1  0.619195  0.380805        0      61.919505
1      pclass2  0.429603  0.570397        0      42.960289
2      pclass3  0.255289  0.744711        0      25.528914
3   Sex female  0.727468  0.272532        0      72.746781
4     Sex male  0.190985  0.809015        0      19.098458
5       parch0  0.335329  0.664671        0      33.532934
6       parch2  0.588235  0.411765        0      58.823529
7       parch1  0.504425  0.495575        0      50.442478
8       parch4  0.625000  0.375000        0      62.500000
9       parch3  0.166667  0.833333        0      16.666667
10      parch5  0.166667  0.833333        0      16.666667
11      parch6  0.000000  1.000000        0       0.000000
12      parch9  0.000000  1.000000        0       0.000000
13      Port C  0.555556  0.444444        2      55.555556
14      Port Q  0.357724  0.642276        2      35.772358
15      Port S  0.332604  0.667396        2      33.260394

## 2. Implement decision tree

### 2.1 Prepare data 

In [183]:
import statsmodels.formula.api as smf

In [364]:
smf.ols('d.survived~d.sex*d.pclass*d.parch*d.age',data=d).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             d.survived   R-squared:                       0.406
Model:                            OLS   Adj. R-squared:                  0.398
Method:                 Least Squares   F-statistic:                     46.99
Date:                Mon, 04 Jun 2018   Prob (F-statistic):          1.74e-105
Time:                        16:35:05   Log-Likelihood:                -468.60
No. Observations:                1046   AIC:                             969.2
Df Residuals:                    1030   BIC:                             1048.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                1.4484      0.171      8.449      0.000       1.112       1.785
d.sex[T.male]                           -1.0906      0.230     -4.739      0.000      -1.542      -0.639
d.pclass                                -0.2886      0.075     -3.861      0.000      -0.435      -0.142
d.sex[T.male]:d.pclass                   0.2629      0.097      2.721      0.007       0.073       0.452
d.parch                                 -0.1181      0.147     -0.806      0.420      -0.406       0.169
d.sex[T.male]:d.parch                    0.6945      0.203      3.429      0.001       0.297       1.092
d.pclass:d.parch                         0.0388      0.062      0.625      0.532      -0.083       0.161
d.sex[T.male]:d.pclass:d.parch          -0.2146      0.084     -2.563      0.011      -0.379      -0.050
d.age                                   -0.0055      0.005     -1.213      0.225      -0.014       0.003
d.sex[T.male]:d.age                      0.0060      0.006      1.018      0.309      -0.006       0.017
d.pclass:d.age                           0.0012      0.002      0.515      0.607      -0.003       0.006
d.sex[T.male]:d.pclass:d.age            -0.0030      0.003     -1.090      0.276      -0.009       0.002
d.parch:d.age                            0.0049      0.004      1.295      0.196      -0.003       0.012
d.sex[T.male]:d.parch:d.age             -0.0163      0.005     -3.262      0.001      -0.026      -0.006
d.pclass:d.parch:d.age                  -0.0020      0.002     -1.209      0.227      -0.005       0.001
d.sex[T.male]:d.pclass:d.parch:d.age     0.0051      0.002      2.325      0.020       0.001       0.009
==============================================================================
Omnibus:                       74.259   Durbin-Watson:                   1.889
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               89.221
Skew:                           0.714   Prob(JB):                     4.23e-20
Kurtosis:                       3.095   Cond. No.                     3.09e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.09e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### 1. Pclass, Sex, Parch , Age are the variables selected for the further use as they have high correlation and influence on the survival rate of the person.

#### 2.Splitting data

In [ ]:
from sklearn.model_selection import train_test_split

In [274]:
x_train,x_test,y_train,y_test=train_test_split(d[['pclass','sex','parch','embarked','age']],d['survived'],test_size=0.2)

### 2.2 Implement the decision tree

In [345]:
#1

eroot=-(d2.survived.mean())*np.log2(d2.survived.mean())-(1-d2.survived.mean())*np.log2(1-d2.survived.mean())
eroot

0.9511183912871488

In [276]:
#2
#The different age values in the training set

np.sort(x_train.age.unique())

array([ 0.1667,  0.3333,  0.4167,  0.6667,  0.75  ,  0.8333,  0.9167,
        1.    ,  2.    ,  3.    ,  4.    ,  5.    ,  6.    ,  7.    ,
        8.    ,  9.    , 10.    , 11.    , 12.    , 13.    , 14.    ,
       15.    , 16.    , 17.    , 18.    , 18.5   , 19.    , 20.    ,
       20.5   , 21.    , 22.    , 23.    , 23.5   , 24.    , 25.    ,
       26.    , 26.5   , 27.    , 28.    , 28.5   , 29.    , 30.    ,
       30.5   , 31.    , 32.    , 32.5   , 33.    , 34.    , 34.5   ,
       35.    , 36.    , 36.5   , 37.    , 38.    , 38.5   , 39.    ,
       40.    , 40.5   , 41.    , 42.    , 43.    , 44.    , 45.    ,
       45.5   , 46.    , 47.    , 48.    , 49.    , 50.    , 51.    ,
       52.    , 53.    , 54.    , 55.    , 55.5   , 56.    , 57.    ,
       58.    , 59.    , 60.    , 61.    , 62.    , 63.    , 64.    ,
       65.    , 66.    , 70.    , 70.5   , 71.    , 74.    ,     nan])

In [277]:
vect=pd.Series(np.sort(x_train.age.unique()))

In [278]:
len(x_train.age.unique())-1 #subtracting 1 to remove nulls

90

There are 90 different age values in the training set.

In [334]:
def entropy(x,y):
    if y==0:
        a=1
    elif x==0:
        a=0
    else:
        a=-(x/(x+y))*np.log2(x/(x+y))-(y/(x+y))*np.log2(y/(x+y))
    return a

In [336]:
entropy(0,0)

1

In [280]:
entropy(2,8)

0.7219280948873623

In [281]:
d2=pd.concat([x_train,y_train],axis=1)
len(d2)

1047

In [282]:
#3
a11=x_train[x_train.age<=0.3333]
a11s=len(d2[(d2.age<=0.3333) & (d2.survived==1)])
a11d=len(d2[(d2.age<=0.3333) & (d2.survived==0)])

In [283]:
a11

pclass     sex  parch embarked     age
747       3    male      2        S  0.3333
763       3  female      2        S  0.1667

In [284]:
a12=x_train[x_train.age>0.3333]
a12s=len(d2[(d2.age>0.3333) & (d2.survived==1)])
a12d=len(d2[(d2.age>0.3333) & (d2.survived==0)])

In [287]:
e11=len(a11)/len(d2)*entropy(a11s,a11d)
e11

0.0019102196752626551

In [288]:
e12=len(a12)/len(d2)*entropy(a12s,a12d)
e12

0.7657790102573567

In [290]:
#4
eage1=e11+e12
eage1

0.7676892299326193

In [338]:
#5

#Entropy values of split with respect to age
entropy_age=[]
for i in vect.iloc[:-1]:
    print('For the age group: ',i)
    age1=x_train[x_train.age<=i]
    as1=len(d2[(d2.age<=i) & (d2.survived==1)])
    ad1=len(d2[(d2.age<=i) & (d2.survived==0)])
    age2=x_train[x_train.age>i]
    as2=len(d2[(d2.age>i) & (d2.survived==1)])
    ad2=len(d2[(d2.age>i) & (d2.survived==0)])
    e=(len(age1)/len(d2))*entropy(as1,ad1)+(len(age2)/len(d2))*entropy(as2,ad2)
    entropy_age.append(e)
    print('Entropy is ', e)

For the age group:  0.1667
Entropy is  0.7674363371509733
For the age group:  0.3333
Entropy is  0.7676892299326193
For the age group:  0.4167
Entropy is  0.7671447192227802
For the age group:  0.6667
Entropy is  0.7663449087203313
For the age group:  0.75
Entropy is  0.7665390419714864
For the age group:  0.8333
Entropy is  0.7657758050253693
For the age group:  0.9167
Entropy is  0.7640237748035283
For the age group:  1.0
Entropy is  0.7611520376035287
For the age group:  2.0
Entropy is  0.7653562171495877
For the age group:  3.0
Entropy is  0.7642224859972976
For the age group:  4.0
Entropy is  0.764133165052945
For the age group:  5.0
Entropy is  0.762994331713255
For the age group:  6.0
Entropy is  0.7629392029961971
For the age group:  7.0
Entropy is  0.7628693750419333
For the age group:  8.0
Entropy is  0.7616744670759902
For the age group:  9.0
Entropy is  0.7609937171738853
For the age group:  10.0
Entropy is  0.7620196714293801
For the age group:  11.0
Entropy is  0.76287329

In [352]:
print('The min entropy is',np.min(entropy_age),'for the age group 15')

The min entropy is 0.7593471485645501 for the age group 15


In [354]:
#6

#Entropy values of split with respect to sex
sex1=x_train[x_train.sex=='male']
ss1=len(d2[(d2.sex=='male') & (d2.survived==1)])
sd1=len(d2[(d2.sex=='male') & (d2.survived==0)])
sex2=x_train[x_train.sex=='female']
ss2=len(d2[(d2.sex=='female') & (d2.survived==1)])
sd2=len(d2[(d2.sex=='female') & (d2.survived==0)])
e2=(len(sex1)/len(d2))*entropy(ss1,sd1)+(len(sex2)/len(d2))*entropy(ss2,sd2)
print('Entropy with respect to sex is', e2)

Entropy with respect to sex is 0.7381271184233056


In [351]:
vect2=pd.Series(np.sort(x_train.pclass.unique()))
vect2

0    1
1    2
2    3
dtype: int64

In [362]:
#Entropy values of split with respect to pclass
entropy_pclass=[]
for i in vect2:
    print('For the pclass group: ',i)
    pclass1=x_train[x_train.pclass<=i]
    ps1=len(d2[(d2.pclass<=i) & (d2.survived==1)])
    pd1=len(d2[(d2.pclass<=i) & (d2.survived==0)])
    pclass2=x_train[x_train.pclass>i]
    ps2=len(d2[(d2.pclass>i) & (d2.survived==1)])
    pd2=len(d2[(d2.pclass>i) & (d2.survived==0)])
    e3=(len(pclass1)/len(d2))*entropy(ps1,pd1)+(len(pclass2)/len(d2))*entropy(ps2,pd2)
    entropy_pclass.append(e3)
    print('Entropy is ', e3)

For the pclass group:  1
Entropy is  0.8897023777301396
For the pclass group:  2
Entropy is  0.8858709386631788
For the pclass group:  3
Entropy is  0.9511183912871487


In [361]:
vect3=pd.Series(np.sort(x_train.parch.unique()))
vect3

0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    9
dtype: int64

In [363]:
#Entropy values of split with respect to parch
entropy_parch=[]
for i in vect3:
    print('For the parch number: ',i)
    parch1=x_train[x_train.parch<=i]
    pas1=len(d2[(d2.parch<=i) & (d2.survived==1)])
    pad1=len(d2[(d2.parch<=i) & (d2.survived==0)])
    parch2=x_train[x_train.parch>i]
    pas2=len(d2[(d2.parch>i) & (d2.survived==1)])
    pad2=len(d2[(d2.parch>i) & (d2.survived==0)])
    e4=(len(parch1)/len(d2))*entropy(pas1,pad1)+(len(parch2)/len(d2))*entropy(pas2,pad2)
    entropy_parch.append(e4)
    print('Entropy is ', e4)

For the parch number:  0
Entropy is  0.9328668279020409
For the parch number:  1
Entropy is  0.9487310213026103
For the parch number:  2
Entropy is  0.9499325981543705
For the parch number:  3
Entropy is  0.9463713433241798
For the parch number:  4
Entropy is  0.9459900033066303
For the parch number:  5
Entropy is  0.9492011104306504
For the parch number:  6
Entropy is  0.9498409821217193
For the parch number:  9
Entropy is  0.9511183912871487


In [368]:
#7

#sThe lowest entropy is with the split at sex

#Highest entropy gain with respect to sex
Entropy_gain=eroot-e2
Entropy_gain

0.21299127286384323

#### Thus, after comparing the values, we found out that sex is the best feature for the first split as it has the lowest entropy thereby giving the highest information gain.

In [374]:
#Percentage of survived and non-survived in first branch

print("The Survived percentage in 1st branch is : ", len(d2[(d2.sex=='female') & (d2.survived==1)])/len(d2[d2.sex=='female'])*100)
print("The Non-Survived percentage in 1st branch is : ", (1-len(d2[(d2.sex=='female') & (d2.survived==1)])/len(d2[d2.sex=='female']))*100)

The Survived percentage in 1st branch is :  71.80851063829788
The Non-Survived percentage in 1st branch is :  28.191489361702125


In [375]:
#Percentage of survived and non-survived in second branch

print("The Survived percentage in 2nd branch is : ", len(d2[(d2.sex=='male') & (d2.survived==1)])/len(d2[d2.sex=='male'])*100)
print("The Non-Survived percentage in 2nd branch is : ", (1-len(d2[(d2.sex=='male') & (d2.survived==1)])/len(d2[d2.sex=='male']))*100)

The Survived percentage in 2nd branch is :  17.58569299552906
The Non-Survived percentage in 2nd branch is :  82.41430700447094


In [379]:
#1st branch data
d21=d2[d2.sex=='female']

In [380]:
#2nd branch data
d22=d2[d2.sex=='male']i

##  1st branch split

In [408]:
vage=pd.Series(np.sort(d21.age.unique()))

In [409]:
#1st branch split with respect to age

entropy_age2=[]
for i in vage.iloc[:-1]:
    print('For the age group: ',i)
    age1=d21[d21.age<=i]
    as1=len(d21[(d21.age<=i) & (d21.survived==1)])
    ad1=len(d21[(d21.age<=i) & (d21.survived==0)])
    age2=d21[d21.age>i]
    as2=len(d21[(d21.age>i) & (d21.survived==1)])
    ad2=len(d21[(d21.age>i) & (d21.survived==0)])
    e21=(len(age1)/len(d21))*entropy(as1,ad1)+(len(age2)/len(d21))*entropy(as2,ad2)
    entropy_age2.append(e21)
    print('Entropy is ', e21)

For the age group:  0.1667
Entropy is  0.677781537132748
For the age group:  0.75
Entropy is  0.6793103319129103
For the age group:  0.9167
Entropy is  0.6808348615793572
For the age group:  1.0
Entropy is  0.6752389759570171
For the age group:  2.0
Entropy is  0.6710072211509404
For the age group:  3.0
Entropy is  0.6672046264036683
For the age group:  4.0
Entropy is  0.671082805938947
For the age group:  5.0
Entropy is  0.673568671961989
For the age group:  6.0
Entropy is  0.6718990358920011
For the age group:  7.0
Entropy is  0.6726815835093667
For the age group:  8.0
Entropy is  0.6724849602877297
For the age group:  9.0
Entropy is  0.6696819227797981
For the age group:  10.0
Entropy is  0.6675824921204376
For the age group:  11.0
Entropy is  0.6652669790661709
For the age group:  12.0
Entropy is  0.66640950940632
For the age group:  13.0
Entropy is  0.6684313409910172
For the age group:  14.0
Entropy is  0.6682462831213627
For the age group:  15.0
Entropy is  0.6719606940982643
Fo

In [410]:
print('The min entropy is',np.min(entropy_age2),'for the age group 50')

The min entropy is 0.6612231330147224 for the age group 50


In [395]:
vage2=pd.Series(np.sort(d21.pclass.unique()))
vage2

0    1
1    2
2    3
dtype: int64

In [396]:
#1st branch split with respect to pclass
entropy_pclass2=[]
for i in vage2:
    print('For the pclass group: ',i)
    pclass1=d21[d21.pclass<=i]
    ps1=len(d21[(d21.pclass<=i) & (d21.survived==1)])
    pd1=len(d21[(d21.pclass<=i) & (d21.survived==0)])
    pclass2=d21[d21.pclass>i]
    ps2=len(d21[(d21.pclass>i) & (d21.survived==1)])
    pd2=len(d21[(d21.pclass>i) & (d21.survived==0)])
    e22=(len(pclass1)/len(d21))*entropy(ps1,pd1)+(len(pclass2)/len(d21))*entropy(ps2,pd2)
    entropy_pclass2.append(e22)
    print('Entropy is ', e22)

For the pclass group:  1
Entropy is  0.7438189068810472
For the pclass group:  2
Entropy is  0.6460214283029786
For the pclass group:  3
Entropy is  0.8580468852217238


In [397]:
print('The min entropy is',np.min(entropy_pclass2),'for pclass group 2')

The min entropy is 0.6460214283029786 for pclass group 2


In [398]:
vage3=pd.Series(np.sort(d21.parch.unique()))
vage3

0    0
1    1
2    2
3    3
4    4
5    5
6    6
7    9
dtype: int64

In [399]:
#Entropy values of 1st split with respect to parch
entropy_parch2=[]
for i in vage3:
    print('For the parch number: ',i)
    parch1=d21[d21.parch<=i]
    pas1=len(d21[(d21.parch<=i) & (d21.survived==1)])
    pad1=len(d21[(d21.parch<=i) & (d21.survived==0)])
    parch2=d21[d21.parch>i]
    pas2=len(d21[(d21.parch>i) & (d21.survived==1)])
    pad2=len(d21[(d21.parch>i) & (d21.survived==0)])
    e23=(len(parch1)/len(d21))*entropy(pas1,pad1)+(len(parch2)/len(d21))*entropy(pas2,pad2)
    entropy_parch2.append(e23)
    print('Entropy is ', e23)

For the parch number:  0
Entropy is  0.85367272573818
For the parch number:  1
Entropy is  0.8404940880797093
For the parch number:  2
Entropy is  0.8423053277897845
For the parch number:  3
Entropy is  0.82917789775377
For the parch number:  4
Entropy is  0.8334244515017617
For the parch number:  5
Entropy is  0.8482781545414313
For the parch number:  6
Entropy is  0.8531756754256622
For the parch number:  9
Entropy is  0.8580468852217238


#### The minimum entropy for 2nd split is with respect to P class group 2 as the entropy is least with the value of 0.6460 thereby getting the maximum information gain.

In [430]:
#1st branch data
de21=d21[d21.pclass<=2]

In [432]:
#2nd branch data 
de22=d21[d21.pclass>2]

In [433]:
#Percentage of survived and non-survived after splitting first branch with respect to pclass group 2

print("The Survived percentage in 1st branch is : ", len(d21[(d21.pclass<=2) & (d21.survived==1)])/len(d21[d21.pclass<=2])*100)
print("The Non- Survived percentage in 1st branch is : ",100- len(d21[(d21.pclass<=2) & (d21.survived==1)])/len(d21[d21.pclass<=2])*100)

The Survived percentage in 1st branch is :  93.93939393939394
The Non- Survived percentage in 1st branch is :  6.060606060606062


In [435]:
#Percentage of survived and non-survived after splitting first branch with respedct to pclass group 2

print("The Survived percentage in 2nd branch is : ", len(d21[(d21.pclass>2) & (d21.survived==1)])/len(d21[d21.pclass>2])*100)
print("The Non- Survived percentage in 2nd branch is : ",100- len(d21[(d21.pclass>2) & (d21.survived==1)])/len(d21[d21.pclass>2])*100)

The Survived percentage in 2nd branch is :  47.19101123595505
The Non- Survived percentage in 2nd branch is :  52.80898876404495


### 2nd split branch 1

In [436]:
va=pd.Series(np.sort(de21.age.unique()))

In [437]:
#1st branch 2nd split with respect to age

entropy_age3=[]
for i in va.iloc[:-1]:
    print('For the age group: ',i)
    age1=de21[de21.age<=i]
    as1=len(de21[(de21.age<=i) & (de21.survived==1)])
    ad1=len(de21[(de21.age<=i) & (de21.survived==0)])
    age2=de21[de21.age>i]
    as2=len(de21[(de21.age>i) & (de21.survived==1)])
    ad2=len(de21[(de21.age>i) & (de21.survived==0)])
    ee21=(len(age1)/len(de21))*entropy(as1,ad1)+(len(age2)/len(de21))*entropy(as2,ad2)
    entropy_age3.append(ee21)
    print('Entropy is ', ee21)

For the age group:  0.9167
Entropy is  0.30943356813346473
For the age group:  1.0
Entropy is  0.3140386571354987
For the age group:  2.0
Entropy is  0.2990314069383298
For the age group:  3.0
Entropy is  0.30046184966143974
For the age group:  4.0
Entropy is  0.3015160560484249
For the age group:  7.0
Entropy is  0.3023208419047213
For the age group:  8.0
Entropy is  0.3034407462085569
For the age group:  12.0
Entropy is  0.303829497517894
For the age group:  13.0
Entropy is  0.30413436961534246
For the age group:  14.0
Entropy is  0.3043702044185248
For the age group:  15.0
Entropy is  0.3046765952807391
For the age group:  16.0
Entropy is  0.3047623556459562
For the age group:  17.0
Entropy is  0.30481179604701447
For the age group:  18.0
Entropy is  0.3040430064515631
For the age group:  19.0
Entropy is  0.30468593182102377
For the age group:  20.0
Entropy is  0.30481769958941574
For the age group:  21.0
Entropy is  0.3047025637500958
For the age group:  22.0
Entropy is  0.30472422

In [439]:
print('The min entropy is',np.min(entropy_age3),'for the age group 62')

The min entropy is 0.2969638292429368 for the age group 62


In [440]:
va2=pd.Series(np.sort(de21.parch.unique()))
va2

0    0
1    1
2    2
3    3
4    4
dtype: int64

In [441]:
#Entropy values of 2nd split of 1st branch with respect to parch
entropy_parch3=[]
for i in va2:
    print('For the parch number: ',i)
    parch1=de21[de21.parch<=i]
    pas1=len(de21[(de21.parch<=i) & (de21.survived==1)])
    pad1=len(de21[(de21.parch<=i) & (de21.survived==0)])
    parch2=de21[de21.parch>i]
    pas2=len(de21[(de21.parch>i) & (de21.survived==1)])
    pad2=len(de21[(de21.parch>i) & (de21.survived==0)])
    ee23=(len(parch1)/len(de21))*entropy(pas1,pad1)+(len(parch2)/len(de21))*entropy(pas2,pad2)
    entropy_parch3.append(ee23)
    print('Entropy is ', ee23)

For the parch number:  0
Entropy is  0.327547015538825
For the parch number:  1
Entropy is  0.3298106822303092
For the parch number:  2
Entropy is  0.34820665080548574
For the parch number:  3
Entropy is  0.3344398395390631
For the parch number:  4
Entropy is  0.32984607020714635


#### The min entropy for 2nd split is 0.2969 for the age group 62

In [446]:
print("The Survived percentage in 1st branch is : ", len(de21[(de21.age<=62) & (de21.survived==1)])/len(de21[de21.age<=62])*100)
print("The Non- Survived percentage in 1st branch is : ",100- len(de21[(de21.age<=62) & (de21.survived==1)])/len(de21[de21.age<=62])*100)

The Survived percentage in 1st branch is :  94.56521739130434
The Non- Survived percentage in 1st branch is :  5.434782608695656


In [447]:
print("The Survived percentage in 2nd branch is : ", len(de21[(de21.age>62) & (de21.survived==1)])/len(de21[de21.age>62])*100)
print("The Non- Survived percentage in 2nd branch is : ",100- len(de21[(de21.age>62) & (de21.survived==1)])/len(de21[de21.age>62])*100)

The Survived percentage in 2nd branch is :  66.66666666666666
The Non- Survived percentage in 2nd branch is :  33.33333333333334


### 2nd split branch 2

In [486]:
#FUNCTION TO CALCULATE THE ENTROPIES
def fentropy(x,y):
    en=[]
    vec=pd.Series(np.sort(y.unique()))
    for i in vec:
        print('For the group number: ',i)
        parch1=x[y<=i]
        pas1=len(x[(y<=i) & (x.survived==1)])
        pad1=len(x[(y<=i) & (x.survived==0)])
        parch2=x[y>i]
        pas2=len(x[(y>i) & (x.survived==1)])
        pad2=len(x[(y>i) & (x.survived==0)])
        ee23=(len(parch1)/len(x))*entropy(pas1,pad1)+(len(parch2)/len(x))*entropy(pas2,pad2)
        en.append(ee23)
        print(ee23)
    print('The lowest entropy is :',min(en))

In [487]:
#Entropy values of 2nd split of 1st branch with respect to parch

fentropy(de22,de22.parch)

For the group number:  0
0.9805370515099496
For the group number:  1
0.9691040490836185
For the group number:  2
0.9700718010630626
For the group number:  3
0.9549608243011968
For the group number:  4
0.971324407369259
For the group number:  5
0.987289895559007
For the group number:  6
0.9925266821248752
For the group number:  9
0.9977221075331743
The lowest entropy is : 0.9549608243011968


In [488]:
fentropy(de22,de22.age)

For the group number:  0.1667
0.691050711841838
For the group number:  0.75
0.6901042520270885
For the group number:  1.0
0.681135467438621
For the group number:  2.0
0.6919075273781777
For the group number:  3.0
0.6903631638699461
For the group number:  4.0
0.6918923087746824
For the group number:  5.0
0.6882381300341884
For the group number:  6.0
0.6899741520665036
For the group number:  8.0
0.6911120912500086
For the group number:  9.0
0.691666351900723
For the group number:  10.0
0.6919119321333549
For the group number:  11.0
0.6917845687643589
For the group number:  13.0
0.6918937150294613
For the group number:  14.0
0.6918627086552253
For the group number:  15.0
0.6892575646125655
For the group number:  16.0
0.6874550679459799
For the group number:  17.0
0.6871141823960905
For the group number:  18.0
0.6878302917081073
For the group number:  18.5
0.6891178527954553
For the group number:  19.0
0.6836444903638519
For the group number:  20.0
0.68848623864541
For the group number:  2

The lowest entropy for age group 27 is 0.6776

## 2nd BRANCH SPLIT

In [503]:
#Splitting on age

fentropy(d22,d22.age)

For the group number:  0.3333
0.5424130674874711
For the group number:  0.4167
0.5418351498012429
For the group number:  0.6667
0.5393838598647402
For the group number:  0.75
0.540789321263149
For the group number:  0.8333
0.5384730702777168
For the group number:  0.9167
0.5358409600973429
For the group number:  1.0
0.5293972395060004
For the group number:  2.0
0.5293206067525348
For the group number:  3.0
0.521073219579655
For the group number:  4.0
0.5260337558816615
For the group number:  5.0
0.5273755214574286
For the group number:  6.0
0.5237293688366471
For the group number:  7.0
0.5238402380152349
For the group number:  8.0
0.5201377373382003
For the group number:  9.0
0.5136790554713536
For the group number:  10.0
0.5150757861216777
For the group number:  11.0
0.51511751991798
For the group number:  12.0
0.5125159778401551
For the group number:  13.0
0.5125098473097195
For the group number:  14.0
0.5149385625351038
For the group number:  15.0
0.5160748532569963
For the group nu

In [492]:
#Splitting on pclass

fentropy(d22,d22.pclass)

For the group number:  1
0.6391971204304095
For the group number:  2
0.6563365471652185
For the group number:  3
0.6709291567001979
The lowest entropy is : 0.6391971204304095


In [505]:
#Splitting on parch 
fentropy(d22,d22.parch)

For the group number:  0
0.6573008900194899
For the group number:  1
0.6695477447978032
For the group number:  2
0.6680013467174685
For the group number:  3
0.6688413281406776
For the group number:  4
0.6696785289439009
For the group number:  5
0.6705129675335227
For the group number:  9
0.6709291567001979
The lowest entropy is : 0.6573008900194899


#### The next split is on Age group 13 as the lowest entropy of 0.5125 was found that group.

In [507]:
d221=d22[d22.age<=13]

In [506]:
d222=d22[d22.age>13]

In [509]:
#Splitting Age group >13 with respect to pclass

fentropy(d222,d222.pclass)

For the group number:  1
0.5920063064102918
For the group number:  2
0.6273273167383945
For the group number:  3
0.6351164767504011
The lowest entropy is : 0.5920063064102918


In [510]:
#Splitting Age group >13 with respect to parch

fentropy(d222,d222.parch)

For the group number:  0
0.6350122143304313
For the group number:  1
0.6347239445039348
For the group number:  2
0.631902824578395
For the group number:  3
0.6329790232025846
For the group number:  4
0.6340502251593907
For the group number:  5
0.6351164767504011
The lowest entropy is : 0.631902824578395


#### The next split for age group >13 is on pclass 1 as the lowest entropy of 0.59200 was found that group.

In [512]:
#Splitting Age group <=13 with respect to pclass

fentropy(d221,d221.pclass)

For the group number:  1
0.9995698112198212
For the group number:  2
0.9282223708830589
For the group number:  3
0.996485989886783
The lowest entropy is : 0.9282223708830589


In [513]:
#Splitting Age group <=13 with respect to parch

fentropy(d221,d221.parch)

For the group number:  0
0.9963142749742503
For the group number:  1
0.9793752078710399
For the group number:  2
0.996485989886783
The lowest entropy is : 0.9793752078710399


#### The next split for age group <=13 is on parch number 1 as the lowest entropy of 0.9793  was found that group.

In [526]:
def fentropy2(col_data, col_name):
    r = pd.DataFrame(columns = ['value', 'entropy', 'weight'])
    len_tot = len(col_data)
    gain = 0
    for i in col_data[col_name].unique():
        x = col_data[col_data[col_name] == i]
        col_weight = len(x)
        e1 = len(x[x['survived'] == 1])
        e2 = len(x[x['survived'] == 0])
        entr = entropy(e1, e2)
        idx = len(r) + 1
        r.loc[idx] = [i,entr,col_weight]
        gain = gain + (entr*col_weight/len_tot)
    
    return(r,gain)

In [528]:
dff, col2 = fentropy2(d2, 'pclass')
print(dff)
print("column entropy is", col2)

   value   entropy  weight
1    3.0  0.790299   565.0
2    1.0  0.957932   258.0
3    2.0  0.981287   224.0
column entropy is 0.8724679590932525


In [529]:
def split(data, cols):
    final_entropy  = pd.DataFrame(columns = ['split_value', 'entropy', 'column'])
    for col in cols:
        for i in data[col].unique():
            data['class'] = np.where(data[col] > i, 1,0)
            
            df, ent = fentropy2(data, 'class')
            idx = len(final_entropy) + 1
            final_entropy.loc[idx] = [i, ent, col]
    return(final_entropy[final_entropy['entropy'] == final_entropy['entropy'].min()])

In [540]:
def pfind(data):
    if(len(data['survived'].unique()) > 0):
        p1 = len(data[data['survived'] == 1])/len(data)*100
        p0 = len(data[data['survived'] == 0])/len(data)*100
    else:   
        if(data['survived'].unique() == 0):
            p0 = 100
            p1 = 0
        else:
            p1 = 100
            p0 = 0

    return(p1, p0)

In [544]:
tree = pd.DataFrame(columns = ['path', 'data_length', 'sur', 'not_sur', 'columnOfSplit', 'thresholdOfSplit'])
path = 'root'
def tree_maker(data, path):
    cols = data.columns.drop(['survived'])
    re = split(data, cols)
    if(np.shape(re)[0]>1):
        return(1)
    per_sur, per_nsur = pfind(data)
    tc = re['column'].item()
    sp = re['split_value'].item()
    data_left = data[ data[tc] > sp]
    data_right = data.drop(index = data_left.index)
    idx = len(tree) + 1
    tree.loc[idx] = [path, len(data),per_sur, per_nsur, tc, sp]
    if(per_sur == 0 or per_nsur == 0):
        idx = len(tree) + 1
        pathleaf = path + 'leaf'
        tree.loc[idx] = [pathleaf, len(data),per_sur, per_nsur, tc, sp]
        return(1)
    else:
        pathl = path + 'Left'
        pathr = path + 'Right'
        tree_maker(data_left, pathl)
        tree_maker(data_right, pathr)  

x = tree_maker(d2f, path)

C:\Users\Prathik\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [539]:
d2f = d2.dropna()

In [537]:
d2f

pclass     sex  parch embarked    age  survived
608        3    male      0        S  30.00         0
629        3    male      5        S  39.00         0
648        3    male      0        C  21.00         0
709        3  female      0        Q  16.00         1
1276       3  female      0        S  31.00         0
1063       3    male      0        S  41.00         0
717        3    male      0        C  18.00         0
676        3    male      0        S  18.00         0
139        1  female      1        S  58.00         1
459        2    male      0        S  42.00         0
873        3    male      0        S  42.00         0
339        2    male      1        S   1.00         1
295        1    male      2        C  17.00         1
949        3    male      0        S  26.00         1
643        3  female      2        S   5.00         1
158        1    male      0        S  55.00         0
1288       3    male      0        S  18.00         0
773        3    male      0        S  42.00         0
735        3  female      2        S  36.00         1
610        3  female      0        S  40.00         0
632        3  female      5        S  39.00         0
406        2    male      0        S  40.00         0
723        3    male      0        Q  31.00         0
1104       3    male      1        S  16.00         0
395        2  female      1        S  18.00         1
612        3  female      1        S  18.00         1
226        1    male      0        S  29.00         0
834        3    male      0        S  32.00         0
501        2  female      1        S  13.00         1
853        3    male      0        S  25.00         0
...      ...     ...    ...      ...    ...       ...
165        1    male      0        S  38.00         1
855        3    male      0        C  11.00         0
555        2    male      0        S  25.00         0
1111       3    male      1        S   0.75         0
450        2    male      0        S  50.00         0
624        3  female      2        S   2.00         0
1227       3  female      0        S  22.00         0
1210       3    male      4        S  40.00         0
1256       3    male      1        C   7.00         1
375        2    male      0        S  28.00         0
1096       3    male      1        S   2.00         0
742        3    male      0        S  45.00         1
700        3    male      0        S  17.00         0
249        1    male      2        C  13.00         1
663        3    male      0        S  26.00         0
577        2  female      0        S  31.00         1
128        1    male      0        S  47.00         0
1226       3    male      0        S  19.00         0
259        1    male      0        S  34.00         1
379        2  female      1        S  31.00         1
78         1  female      2        C  64.00         1
345        2    male      0        S  23.00         0
315        1  female      0        S  21.00         1
1239       3    male      0        S  34.00         0
356        2    male      0        S  25.00         0
229        1  female      0        C  17.00         1
811        3  female      3        S  48.00         0
48         1  female      0        C  53.00         1
358        2  female      0        S  42.00         1
587        2    male      1        S   2.00         1

[826 rows x 6 columns]

In [545]:
tree

path data_length        sur  \
1                                                 root         826  39.830508   
2                                             rootLeft         517  19.148936   
3                                         rootLeftLeft         474  16.033755   
4                                     rootLeftLeftLeft         358  10.614525   
5                                 rootLeftLeftLeftLeft         114   4.385965   
6                            rootLeftLeftLeftLeftRight          84   5.952381   
7                        rootLeftLeftLeftLeftRightLeft           7  28.571429   
8                    rootLeftLeftLeftLeftRightLeftLeft           5  40.000000   
9                       rootLeftLeftLeftLeftRightRight          77   3.896104   
10                  rootLeftLeftLeftLeftRightRightLeft          68   2.941176   
11             rootLeftLeftLeftLeftRightRightLeftRight          29   6.896552   
12         rootLeftLeftLeftLeftRightRightLeftRightLeft          23   4.347826   
13     rootLeftLeftLeftLeftRightRightLeftRightLeftLeft           7  14.285714   
14                 rootLeftLeftLeftLeftRightRightRight           9  11.111111   
15             rootLeftLeftLeftLeftRightRightRightLeft           2  50.000000   
16                               rootLeftLeftLeftRight         244  13.524590   
17                           rootLeftLeftLeftRightLeft          23  30.434783   
18                       rootLeftLeftLeftRightLeftLeft          19  36.842105   
19                   rootLeftLeftLeftRightLeftLeftLeft          13  46.153846   
20                  rootLeftLeftLeftRightLeftLeftRight           6  16.666667   
21             rootLeftLeftLeftRightLeftLeftRightRight           2  50.000000   
22                          rootLeftLeftLeftRightRight         221  11.764706   
23                      rootLeftLeftLeftRightRightLeft         193  10.362694   
24                  rootLeftLeftLeftRightRightLeftLeft          34   2.941176   
25              rootLeftLeftLeftRightRightLeftLeftLeft          11   9.090909   
26          rootLeftLeftLeftRightRightLeftLeftLeftLeft           5  20.000000   
27                 rootLeftLeftLeftRightRightLeftRight         159  11.949686   
28             rootLeftLeftLeftRightRightLeftRightLeft          53  16.981132   
29        rootLeftLeftLeftRightRightLeftRightLeftRight          51  15.686275   
30    rootLeftLeftLeftRightRightLeftRightLeftRightLeft          34  20.588235   
..                                                 ...         ...        ...   
87   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...          31  48.387097   
88   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...          22  40.909091   
89   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...          14  50.000000   
90   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           9  33.333333   
91   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           5  60.000000   
92   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           8  25.000000   
93   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           9  66.666667   
94   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           2  50.000000   
95   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           7  71.428571   
96   rootRightLeftLeftRightLeftLeftLeftLeftLeftLeft...           7  14.285714   
97     rootRightLeftLeftRightLeftLeftLeftLeftLeftRight          15  66.666667   
98         rootRightLeftLeftRightLeftLeftLeftLeftRight          18  27.777778   
99    rootRightLeftLeftRightLeftLeftLeftLeftRightRight          11  45.454545   
100            rootRightLeftLeftRightLeftLeftLeftRight           7  71.428571   
101        rootRightLeftLeftRightLeftLeftLeftRightLeft           2  50.000000   
102                        rootRightLeftLeftRightRight           9  22.222222   
103                                 rootRightLeftRight          17  82.352941   
104                             rootRightLeftRightLeft           2  50.000000   


In [562]:
#Test data 

d3=pd.concat([x_test,y_test],axis=1)
len(d3)
d3.head()

262

pclass     sex  parch embarked    age  survived
657        3  female      1        C   0.75         1
573        2  female      0        S  27.00         1
1120       3    male      0        S  25.00         1
336        2    male      0        S  32.00         1
828        3  female      2        S  10.00         0

In [551]:
path = 'root'
def predict_class(row, path):
    rules = tree[tree['path'] == path]
    #print(rules)
    if(len(rules.index) == 0):
        path = path[:-1]
        rules = tree[tree['path'] == path]
        x = np.where(rules['sur'].item() > rules['not_sur'].item(), 1, 0)
        return(x.item())
    else:
        if(row[rules['columnOfSplit'].item()].item() > rules['thresholdOfSplit'].item()):
            path = path + 'Left'
            return(predict_class(row, path))
        else:
            path = path + 'Right'
            return(predict_class(row, path))    

In [564]:
path = 'root'
test = pd.DataFrame([[2,1,29.00,'C',9,2]], columns = ['pclass', 'sex', 'parch','embarked','age','survived'])
test

pclass  sex  parch embarked  age  survived
0       2    1   29.0        C    9         2

In [565]:
tree[tree['path'] == 'rootLeftRightLeft']

path data_length        sur    not_sur columnOfSplit  \
75  rootLeftRightLeft          29  31.034483  68.965517      embarked   

   thresholdOfSplit  
75                C

In [ ]:
r2 = list()
for elem in d3.iterrows():
    temp = pd.DataFrame(list(elem)[1:], columns = ['pclass', 'sex', 'parch','embarked','age','survived'])
    path = 'root'
    t = predict_class(temp, path)
    r2.append(t)

len(r2)

In [568]:
def accuracy_finder(data):
    tp = 0
    fp = 0
    tn = 0
    fn = 0
    tp = np.where(data['survived'] == data['predictions'], np.where(data['predictions']==1,1,0),0)
    fp = np.where(data['survived'] != data['predictions'], np.where(data['predictions']==1,1,0),0)
    tn = np.where(data['survived'] == data['predictions'], np.where(data['predictions']==0,1,0),0)
    fn = np.where(data['survived'] != data['predictions'], np.where(data['predictions']==0,1,0),0)
    precision = tp.sum()/(tp.sum()+fp.sum())
    recall = tp.sum()/(tp.sum()+fn.sum())
    fsc = 2/((1/precision) + (1/recall))
    accuracy = (tp.sum()+tn.sum())/(tp.sum()+tn.sum()+fp.sum()+fn.sum())
    return(accuracy,fsc,recall,precision)


This is the decision tree classifier which shows the path from the root node to the leaf of the class.

## 3.Bagging

The number of bags is 5 

In [ ]:
ntree = pd.DataFrame(columns = ['path', 'data_length', 'sur', 'not_sur', 'columnOfSplit', 'thresholdOfSplit'])
path = 'root'
def ntree_maker(data, path):
    cols = data.columns.drop(['survived'])
    re = split(data, cols)
    if(np.shape(re)[0]>1):
        return(1)
    per_sur, per_nsur = pfind(data)
    tc = re['column'].item()
    sp = re['split_value'].item()
    data_left = data[ data[tc] > sp]
    data_right = data.drop(index = data_left.index)
    idx = len(ntree) + 1
    ntree.loc[idx] = [path, len(data),per_sur, per_nsur, tc, sp]
    if(per_sur == 0 or per_nsur == 0):
        idx = len(ntree) + 1
        pathleaf = path + 'leaf'
        ntree.loc[idx] = [pathleaf, len(data),per_sur, per_nsur, tc, sp]
        return(1)
    else:
        pathl = path + 'Left'
        pathr = path + 'Right'
        ntree_maker(data_left, pathl)
        ntree_maker(data_right, pathr)  